util/src/main/java/org/cryptomator/util/crypto/HubDeviceCryptor.java

```diff
@@ -0,0 +1,219 @@
+		try {
+			var factory = KeyFactory.getInstance(EC_ALG);
+			var privateKey = factory.generatePrivate(new PKCS8EncodedKeySpec(encoded));
+			if (privateKey instanceof ECPrivateKey ecPrivateKey) {
+				return ecPrivateKey;
+			} else {
+				throw new IllegalStateException(EC_ALG + " key factory not generating ECPrivateKeys");
+			}
+		} catch (NoSuchAlgorithmException e) {
+			throw new IllegalStateException(EC_ALG + " not supported");
+		} catch (InvalidKeySpecException e) {
+			throw new KeyDecodeFailedException(e);
+		}
+	}
+
+	private static <T> T readKey(JWEObject jwe, String keyField, Function<byte[], T> rawKeyFactory) throws MasterkeyLoadingFailedException {
+		Preconditions.checkArgument(jwe.getState() == JWEObject.State.DECRYPTED);
+		var fields = jwe.getPayload().toJSONObject();
+		if (fields == null) {
+			Timber.tag("HubDeviceCryptor").e("Expected JWE payload to be JSON: " + jwe.getPayload());
+			throw new MasterkeyLoadingFailedException("Expected JWE payload to be JSON");
+		}
+		var keyBytes = new byte[0];
+		try {
+			if (fields.get(keyField) instanceof String key) {
+				keyBytes = Base64.getDecoder().decode(key);
+				return rawKeyFactory.apply(keyBytes);
+			} else {
+				throw new IllegalArgumentException("JWE payload doesn't contain field " + keyField);
+			}
+		} catch (IllegalArgumentException | KeyDecodeFailedException e) {
+			Timber.tag("HubDeviceCryptor").e("Unexpected JWE payload: " + jwe.getPayload());
+			throw new MasterkeyLoadingFailedException("Unexpected JWE payload", e);
+		} finally {
+			Arrays.fill(keyBytes, (byte) 0x00);
+		}
+	}
+
+	private static JWEObject encryptKey(Key key, ECPublicKey userKey) {
+		try {
+			var encodedVaultKey = Base64.getEncoder().encodeToString(key.getEncoded());
+			var keyGen = new ECKeyGenerator(Curve.P_384);
+			var ephemeralKeyPair = keyGen.generate();
+			var header = new JWEHeader.Builder(JWEAlgorithm.ECDH_ES, EncryptionMethod.A256GCM).ephemeralPublicKey(ephemeralKeyPair.toPublicJWK()).build();
+			var payload = new Payload(Map.of(JWE_PAYLOAD_KEY_FIELD, encodedVaultKey));
+			var jwe = new JWEObject(header, payload);
+			jwe.encrypt(new ECDHEncrypter(userKey));
+			return jwe;
+		} catch (JOSEException e) {
+			throw new RuntimeException(e);
+		}
+	}
+
+	public static JWEObject encryptUserKey(ECPrivateKey userKey, ECPublicKey deviceKey) {
+		return encryptKey(userKey, deviceKey);
+	}
+
+	public static Masterkey decryptVaultKey(JWEObject jwe, ECPrivateKey privateKey) throws InvalidJweKeyException {
+		try {
+			jwe.decrypt(new ECDHDecrypter(privateKey));
+			return readKey(jwe, JWE_PAYLOAD_KEY_FIELD, Masterkey::new);
+		} catch (JOSEException e) {
+			throw new InvalidJweKeyException(e);
+		}
+	}
+
+	public JWEObject encryptUserKey(JWEObject userKey, String setupCode) {
+		var userPrivateKey = decryptUserKey(userKey, setupCode);
+		var devicePublicKey = getDevicePublicKey();
+		return encryptUserKey(userPrivateKey, devicePublicKey);
+	}
```

Comment:
The name of this method is misleading. It first *decrypts* the `userKey` and then *encrypts* it with the `devicePublicKey`.
This is particularly confusing when you consider that it calls another method with the same name in line 174, which only performs encryption:

https://github.com/cryptomator/android/blob/19cd927e263d1ba1552d4fe004e05b5da9c44d92/util/src/main/java/org/cryptomator/util/crypto/HubDeviceCryptor.java#L158-L160

https://github.com/cryptomator/android/blob/19cd927e263d1ba1552d4fe004e05b5da9c44d92/util/src/main/java/org/cryptomator/util/crypto/HubDeviceCryptor.java#L143-L156